substrate/network-libp2p/src/network_state.rs

```diff
@@ -0,0 +1,796 @@
+				sender.unbounded_send(message)
+					.map_err(|err| {
+						ErrorKind::Io(IoError::new(IoErrorKind::Other, err))
+					})?;
+				Ok(())
+
+			} else {
+				// We are connected to this peer, but not with the current protocol.
+				debug!(target: "sub-libp2p", "Tried to send message to peer {} for which we aren't \
+					connected with the requested protocol", peer_id);
+				return Err(ErrorKind::PeerNotFound.into());
+			}
+		} else {
+			debug!(target: "sub-libp2p", "Tried to send message to invalid peer ID {}", peer_id);
+			return Err(ErrorKind::PeerNotFound.into());
+		}
+	}
+
+	/// Disconnects a peer, if a connection exists (ie. drops the Kademlia controller, and the
+	/// senders that were passed to `accept_custom_proto`).
+	pub fn disconnect_peer(&self, peer_id: PeerId) {
+		let mut connections = self.connections.write();
+		if let Some(peer_info) = connections.info_by_peer.remove(&peer_id) {
+			let old = connections.peer_by_nodeid.remove(&peer_info.id);
+			debug_assert_eq!(old, Some(peer_id));
+		}
+	}
+
+	/// Disconnects all the peers.
+	/// This destroys all the Kademlia controllers and the senders that were passed to
+	/// `accept_custom_proto`.
+	pub fn disconnect_all(&self) {
+		let mut connec = self.connections.write();
+		*connec = Connections {
+			info_by_peer: FnvHashMap::with_capacity_and_hasher(connec.peer_by_nodeid.capacity(), Default::default()),
+			peer_by_nodeid: FnvHashMap::with_capacity_and_hasher(connec.peer_by_nodeid.capacity(), Default::default()),
+		};
+	}
+
+	/// Disables a peer. This adds the peer to the list of disabled peers, and drops any existing
+	/// connections if necessary (ie. drops the sender that was passed to `accept_custom_proto`).
+	pub fn disable_peer(&self, peer_id: PeerId) {
+		// TODO: what do we do if the peer is reserved?
+		let mut connections = self.connections.write();
+		let peer_info = if let Some(peer_info) = connections.info_by_peer.remove(&peer_id) {
+			let old = connections.peer_by_nodeid.remove(&peer_info.id);
+			debug_assert_eq!(old, Some(peer_id));
+			peer_info
+		} else {
+			return;
+		};
+
+		drop(connections);
+		self.disabled_peers.write().insert(peer_info.id.clone());
+	}
+
+	/// Returns true if a peer is disabled.
+	pub fn is_peer_disabled(&self, node_id: &PeerstorePeerId) -> bool {
+		self.disabled_peers.read().contains(&node_id)
+	}
+
+	/// Flushes the caches to the disk.
+	///
+	/// This is done in an atomical way, so that an error doesn't corrupt anything.
+	pub fn flush_caches_to_disk(&self) -> Result<(), IoError> {
+		match self.peerstore {
+			PeersStorage::Memory(_) => Ok(()),
+			PeersStorage::Json(ref json) => {
+				match json.flush() {
+					Ok(()) => {
+						debug!(target: "sub-libp2p", "Flushed JSON peer store to disk");
+						Ok(())
+					}
+					Err(err) => {
+						warn!(target: "sub-libp2p", "Failed to flush changes to JSON \
+													peer store: {}", err);
```

Comment:
way too many tabs, it seems (carry-on lines should be one-indent more than original)